conduit/data/datasets/vision/nico_plus_plus.py

```diff
@@ -152,90 +143,100 @@ def __init__(
-
-    def _get_label_mapping(self, level: Literal["superclass", "subclass"]) -> List[Tuple[str, int]]:
+    def _get_label_mapping(self, level: Literal["y", "a"]) -> List[Tuple[str, int]]:
         """Get a list of all possible (name, numerical value) pairs."""
         return list(
-            self.metadata[[level, f"{level}_le"]]
+            self.metadata[[f"{level}_name", level]]
             .drop_duplicates()
             .itertuples(index=False, name=None)
         )
 
     def _check_unzipped(self) -> bool:
-        return all((self._base_dir / attr).exists() for attr in NicoPPAttr)
+        if not all((self._base_dir / self.subpath / attr).exists() for attr in NicoPPAttr):
+            return False
+        if not (self._base_dir / "dg_label_id_mapping.json").exists():
+            return False
+        return True
 
     def _extract_metadata(self) -> None:
-        """Extract concept/context/superclass information from the image filepaths and it save to csv."""
-        self.logger.info("Extracting metadata.")
-        image_paths: List[Path] = []
-        image_paths.extend(self._base_dir.glob(f"**/*.jpg"))
-        image_paths_str = [str(image.relative_to(self._base_dir)) for image in image_paths]
-        filepaths = pd.Series(image_paths_str)
-        metadata = filepaths.str.split("/", expand=True).rename(
-            columns={0: "subclass", 1: "superclass", 2: "filename"}
+        self.logger.info("Generating metadata for NICO++...")
+        attributes = ["autumn", "dim", "grass", "outdoor", "rock", "water"]  # 6 attrs, 60 labels
+        meta = json.load(open(self._base_dir / "dg_label_id_mapping.json", "r"))
+
+        def _make_balanced_testset(
+            df: pd.DataFrame, *, seed: int, num_samples_val_test: int
+        ) -> pd.DataFrame:
+            # each group has a test set size of (2/3 * num_samples_val_test) and a val set size of
+            # (1/3 * num_samples_val_test); if total samples in original group < num_samples_val_test,
+            # val/test will still be split by 1:2, but no training samples remained
+
+            random.seed(seed)
+            val_set, test_set = [], []
+            for g in pd.unique(df["g"]):
+                df_group = df[df["g"] == g]
+                curr_data = df_group["filename"].values
+                random.shuffle(curr_data)
+                split_size = min(len(curr_data), num_samples_val_test)
+                val_set += list(curr_data[: split_size // 3])
+                test_set += list(curr_data[split_size // 3 : split_size])
+            self.logger.info(f"Val: {len(val_set)}, Test: {len(test_set)}")
+            assert len(set(val_set).intersection(set(test_set))) == 0
+            combined_set = dict(zip(val_set, [NicoPPSplit.VAL.value for _ in range(len(val_set))]))
+            combined_set.update(
+                dict(zip(test_set, [NicoPPSplit.TEST.value for _ in range(len(test_set))]))
```

Comment:
If I understand this code correctly, it is incredibly inefficient, but whatever.